src/tr1/game/objects/general/pickup.c

```diff
@@ -76,9 +88,53 @@ static void M_GetAllAtLaraPos(ITEM *item, ITEM *lara_item)
     }
 }
 
+static void M_DrawPickup(ITEM *const item)
+{
+    M_DrawPickupHighlight(item);
+    Object_DrawPickupItem(item);
+}
+
+static void M_DrawPickupHighlight(ITEM *const item)
+{
+    if (!g_Config.visuals.highlight_pickups || Phase_Get() != PHASE_GAME
+        || !Object_GetObject(O_PICKUP_TWINKLE)->loaded) {
+        return;
+    }
+
+    const int32_t distance = Item_GetDistance(g_LaraItem, &item->pos);
+    if (distance < HIGHLIGHT_DIST_MIN || distance > HIGHLIGHT_DIST_MAX) {
+        return;
+    }
+
+    const int32_t previous_time =
+        item->priv == NULL ? 0 : (int32_t)(intptr_t)item->priv;
+    const double time = Clock_GetHighPrecisionCounter();
+    const double delta = time - previous_time;
+    if (delta < HIGHLIGHT_WAIT_MIN
+        || (delta < HIGHLIGHT_WAIT_MAX && (Random_GetDraw() % 7) != 0)) {
+        return;
+    }
+
+    item->priv = (void *)(intptr_t)(int32_t)time;
+
+    const GAME_OBJECT_ID obj_id =
+        Object_GetCognate(item->object_id, g_ItemToInvObjectMap);
+    const OBJECT *const object = Object_GetObject(obj_id);
+    const FRAME_INFO *const frame = object->frame_base;
+
+    const GAME_VECTOR pos = {
+        .x = item->pos.x + 20 * (Random_GetDraw() - 0x4000) / 0x4000,
+        .y = item->pos.y - ABS(frame->bounds.max.y - frame->bounds.min.y) - 10
+            - 10 * (Random_GetDraw() - 0x4000) / 0x4000,
+        .z = item->pos.z + 20 * (Random_GetDraw() - 0x4000) / 0x4000,
+        .room_num = item->room_num,
+    };
+    PickupTwinkle_Spawn(&pos);
+}
```

Comment:
I don't think we should manipulate object state or spawn new effects in the draw routines, which should be mostly immutable with regard to the objects and items, and instead would opt to handle this in the control (or possibly collision) routine. By doing so, we can eliminate the need for the check for the current phase, and avoid having to reference the clock, by working with logical frames instead.

Reply:
I forgot to touch on this. I had it initially in the collision routine, but because of the min distance checks [here](https://github.com/LostArtefacts/TRX/blob/develop/src/tr1/game/lara/control.c#L132), pickup collision doesn't necessarily run when the item is visible. I think we could introduce a control routine for pickups though as an alternative?

Reply:
Actually, this is trickier as pickup items aren't classed as active so aren't included in the `Item_Control` loop.

Reply:
What if we count them as active?

Reply:
We'd need to activate items on load depending on visibility, monitor triggered items, carried items etc. Not sure if it would impact saves. I can try to investigate tomorrow (I'm running short on time this evening).
Otherwise, potentially a different collision test approach for pickup items could be a more streamlined approach.